src/log_batch.rs

```diff
@@ -392,208 +332,363 @@ impl<M: MessageExt> LogItem<M> {
+        // { u64 len | u64 offset | (compressed) entries | crc32 | item count | entry indexes / commands / kvs | crc32 }
+        //                                                      ^ offset = <offset>
 
-        let checksum = crc32(&vec[8..]);
-        vec.encode_u32_le(checksum).unwrap();
-        let len = vec.len() as u64 - 8;
-        let mut header = len << 8;
-        header |= u64::from(compression_type.to_byte());
-        vec.as_mut_slice().write_u64::<BigEndian>(header).unwrap();
+        let mut buf = Vec::with_capacity(4096);
 
-        let batch_len = (vec.len() - 8) as u64;
-        for item in self.items.iter_mut() {
-            if let LogItemContent::Entries(entries) = &mut item.content {
-                entries.update_compression_type(compression_type, batch_len as u64);
+        // reserve for header
+        buf.encode_u64(0)?;
+        buf.encode_u64(0)?;
+
+        // encode entries first, set entries_index
+        let mut entries_indexes = self.entries_batch.encode_to(&mut buf, &mut 0)?;
+        self.item_batch.set_entries_indexes(&mut entries_indexes)?;
+
+        let compression_type =
+            if compression_threshold > 0 && (buf.len() - HEADER_LEN) > compression_threshold {
+                buf = lz4::encode_block(&buf[HEADER_LEN..], HEADER_LEN, 4);
+                CompressionType::Lz4
+            } else {
+                CompressionType::None
+            };
+        let checksum = crc32(&buf[HEADER_LEN..]);
+        buf.encode_u32_le(checksum)?;
+
+        let offset = buf.len() as u64;
+
+        // encode footer (entries_index / commands / kvs) with checksum
+        self.item_batch.encode_to(&mut buf)?;
+
+        // write header
+        let len = ((buf.len() as u64) << 8) | u64::from(compression_type.to_byte());
+
+        (&mut buf[..HEADER_LEN]).write_u64::<BigEndian>(len)?;
+        (&mut buf[8..HEADER_LEN]).write_u64::<BigEndian>(offset)?;
+
+        // update entries_index
+        for item in self.item_batch.iter_mut() {
+            if let LogItemContent::EntriesIndex(entries_index) = &mut item.content {
+                for ei in entries_index.0.iter_mut() {
+                    ei.compression_type = compression_type;
+                    ei.entries_offset = HEADER_LEN as u64; // need to add batch offset of the file after written.
+                    ei.entries_len = offset as usize - HEADER_LEN;
+                }
             }
         }
-
-        Some(vec)
+        Ok(Some(buf))
     }
 
     // Don't account for compression and varint encoding.
+    // Not meaningful after recovery.
     pub fn approximate_size(&self) -> usize {
-        if self.items.is_empty() {
+        if self.item_batch.is_empty() {
             0
         } else {
-            8 /*len*/ + 8 /*count*/ + self.items_size + 4 /*checksum*/
+            8 /*len*/ + 8 /*offset*/ + 8 /*items count*/
+            + self.content_approximate_size + CHECKSUM_LEN *2
```

Comment:
Looks like the code isn't formatted?

Reply:
Weird again, my formatter doesn't report it. This happens a lot of times these days.